### ISSUE TITLE
[feature request] Adding Pre and Post padding functionalities to pad_sequence function

### ISSUE THREAD

**Issue by srirambandi** [pytorch/pytorch#10536]
I would like to add **pre** and **post** padding functionalities to ````torch.nn.utils.rnn.pad_sequence```` .
These functionalities haven't been included in the [function](https://pytorch.org/docs/stable/_modules/torch/nn/utils/rnn.html#pad_sequence)

It would be greatly beneficial for people using seq models like LSTM to have sequences created with both **pre** and **post** padding the sequences passed into the args. Other libraries like keras offer these functionalities in their APIs like: ````data = pad_sequences(data, maxlen=50, dtype='float32', padding='pre')````

Besides this, we can add another functionality to allow an argument(with optional value) to know the length of the sequences to be padded to. This is currently done with length of ````sequence[0]````(sequences passed in decreasing order of length).

Thanks!

cc @albanD @mruberry @jbschlosser

**Comment by soumith** [comment]
this seems like a good feature to add.

**Comment by zou3519** [comment]
@srirambandi how is pre-padding useful? Could you point me to some specific use cases?

**Comment by srirambandi** [comment]
@zou3519 Pre-padding has the same use case as post-padding(which pad_sequence currently offer), except it is proving more efficient in LSTMs, as I have observed. People also reported this behavior [here](https://discuss.pytorch.org/t/padding-pre-versus-post-lstm-performance/13616), and [here](
https://github.com/enerrio/OpenAI-XOR#results). So, I think a library which offers post-padding as default, should also include pre-padding func.

**Comment by mskoh52** [comment]
I recently had a need for this type of functionality so I made the necessary changes to . But I just now saw that this issue and associated PR exist but haven't been touched for a few years. Is there still any interest in this functionality?

**Comment by dongin1009** [comment]
Previously, discussed from #10536 #61467 

If this issue still is considerable, I can make a pull request. I modified pad_sequence by adding a `pos` argument. This option will be necessary as same as the other ML frameworks support.

I just `flip` and compute the `dim`ensionality of tensor without manipulating the CPP file and tensor slicing.

```    
def pad_sequence(
    sequences: Union[Tensor, List[Tensor]],
    batch_first: bool = False,
    padding_value: float = 0.0,
    pos: str = 'right',
) -> Tensor:

    if pos='right':
        padded_sequence = torch._C._nn.pad_sequence(sequences, batch_first, padding_value)
    elif pos='left':
        sequences = tuple(map(lambda s: s.flip(0), sequences))
        padded_sequence = torch._C._nn.pad_sequence(sequences, batch_first, padding_value)
        _seq_dim = padded_sequence.dim()
        padded_sequence = padded_sequence.flip(-_seq_dim+batch_first)
    else:
        raise ValueError("pos should be either 'right' or 'left', but got {}".format(pos))
    return padded_sequence
```

**Comment by ringohoffman** [comment]
> how is pre-padding useful? Could you point me to some specific use cases?

@zou3519 Inputs for many causal language models like Llama and Mistral are left-padded. Additionally, Hugging Face's [`PreTrainedTokenizerBase.pad`](https://github.com/huggingface/transformers/blob/9d6c0641c4a3c2c5ecf4d49d7609edd5b745d9bc/src/transformers/tokenization_utils_base.py#L3375) scales <ins>very</ins> poorly with the size of the inputs.

![e6c0f4f7-aae8-45ec-adb2-56deefb4bd56](https://github.com/user-attachments/assets/222c7167-09d3-427d-8dee-3c60ba6a74af)

> I just flip and compute the dimensionality of tensor without manipulating the CPP file and tensor slicing.

@dongin1009 while clever, `flip` is inefficient. All `pad_sequence` really boils down to is copying a bunch of tensors into another bigger tensor, so, flipping each input sequence and the entire result is all just wasted computation.

![d50f01f3-b2fc-4f5a-a157-5cbf745c7d44](https://github.com/user-attachments/assets/53e85e4d-eb20-44e5-a6c9-86f32678d211)

In fact, it is about 2x as slow for every size of input I tested.

```python
>>> [left / right for left, right in zip(left_pad_times, right_pad_times)]
[1.8522874124890147,
 1.8125473737715978,
 1.9836044873296188,
 1.899170326682036,
 1.9772474951962087,
 2.26629303381468]
```

I am very interested in reviving #61467.

---

The code:

```python
import random
import time
from typing import Literal

import numpy as np
import torch
import transformers


def pad_sequence(
    sequences: list[torch.Tensor],
    batch_first: bool = False,
    padding_value: int | float | bool = 0.0,
    padding_side: Literal["left", "right"] | str = "left",
) -> torch.Tensor:
    if padding_side == 'right':
        padded_sequence = torch._C._nn.pad_sequence([t.flatten() for t in sequences], batch_first=batch_first, padding_value=padding_value)
    elif padding_side=='left':
        padded_sequence = torch._C._nn.pad_sequence([t.flatten().flip(0) for t in sequences], batch_first=batch_first, padding_value=padding_value)  # pyright: ignore[reportArgumentType]
        padded_sequence = padded_sequence.flip(int(batch_first))
    else:
        raise ValueError(f"padding_side should be either 'right' or 'left', but got {padding_side}")

    return padded_sequence


tokenizer = transformers.AutoTokenizer.from_pretrained("/models/mistralai/Mistral-7B-Instruct-v0.2")
tokenizer.pad_token_id = tokenizer.eos_token_id
assert tokenizer.padding_side == "left"


sequence_lengths: list[int] = []
left_pad_times: list[float] = []
left_pad_times_std: list[float] = []
right_pad_times: list[float] = []
right_pad_times_std: list[float] = []
tokenizer_pad_times: list[float] = []
tokenizer_pad_times_std: list[float] = []
RUNS_PER_LOOP: int = 10


for i in range(1, 7):  # takes about a minute to run
    sequence_length = 10 ** i
    sequence_lengths.append(sequence_length)

    sequences = [torch.randint(1, 10, (1, random.randint(1, sequence_length))) for _ in range(64)]
    huggingface_inputs = [{"input_ids": seq.flatten()} for seq in sequences]

    torch_padded_sequence = pad_sequence(sequences, padding_value=tokenizer.pad_token_id, batch_first=True, padding_side="left")
    huggingface_padded_sequence = tokenizer.pad(huggingface_inputs, return_tensors="pt", return_attention_mask=False)["input_ids"]
    assert torch_padded_sequence.equal(huggingface_padded_sequence)

    inner_left_pad_times: list[float] = []
    inner_right_pad_times: list[float] = []
    inner_tokenizer_pad_times: list[float] = []

    for _ in range(RUNS_PER_LOOP):

        start = time.perf_counter()
        pad_sequence(sequences, batch_first=True, padding_side="left")
        end = time.perf_counter()
        inner_left_pad_times.append(end - start)

        start = time.perf_counter()
        pad_sequence(sequences, batch_first=True, padding_side="right")
        end = time.perf_counter()
        inner_right_pad_times.append(end - start)

        start = time.perf_counter()
        tokenizer.pad(huggingface_inputs, return_tensors="pt", return_attention_mask=False)
        end = time.perf_counter()
        inner_tokenizer_pad_times.append(end - start)
    
    left_pad_times.append(sum(inner_left_pad_times) / len(inner_left_pad_times))
    left_pad_times_std.append(np.std(inner_left_pad_times))

    right_pad_times.append(sum(inner_right_pad_times) / len(inner_right_pad_times))
    right_pad_times_std.append(np.std(inner_right_pad_times))

    tokenizer_pad_times.append(sum(inner_tokenizer_pad_times) / len(inner_tokenizer_pad_times))
    tokenizer_pad_times_std.append(np.std(inner_tokenizer_pad_times))

    print(f"Sequence Length: {sequence_length}, Left Pad Time: {left_pad_times[-1]}, Right Pad Time: {right_pad_times[-1]}, Tokenizer Pad Time: {tokenizer_pad_times[-1]}")

import matplotlib.pyplot as plt

plt.plot(sequence_lengths, left_pad_times, label="pad_sequence left (2 flips)")
plt.scatter(sequence_lengths, left_pad_times)
plt.errorbar(sequence_lengths, left_pad_times, yerr=left_pad_times_std, linestyle='None', marker='^')

plt.plot(sequence_lengths, right_pad_times, label="pad_sequence right")
plt.scatter(sequence_lengths, right_pad_times)
plt.errorbar(sequence_lengths, right_pad_times, yerr=right_pad_times_std, linestyle='None', marker='^')

plt.plot(sequence_lengths, tokenizer_pad_times, label="Hugging Face tokenizer")
plt.scatter(sequence_lengths, tokenizer_pad_times)
plt.errorbar(sequence_lengths, tokenizer_pad_times, yerr=tokenizer_pad_times_std, linestyle='None', marker='^')

plt.xlabel("Sequence Length")
plt.ylabel("Time (s)")
plt.legend(loc="upper right")
```
